File: java/leetcode/P113_PathSumII.java
package leetcode;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import tools.TreeNode辅助.TreeNode;

public class P113_PathSumII {
	public static void main(String[] args) {
		
	}
	/*
	 * 	AC
	 * 	3 ms
	 */
	static class Solution {
		List<List<Integer>> ans = new LinkedList<>();
		ArrayList<Integer> path = new ArrayList<>();
	    public List<List<Integer>> pathSum(TreeNode root, int sum) {
	    	if (root == null) {
	    		return ans;
	    	}
	    	search(root, sum, 0);
	        return ans;
	    }
		private void search(TreeNode root, int sum, int now) {
			if (root.left == null && root.right == null && now + root.val == sum) {
				List<Integer> temp = new LinkedList<>();
				for (int i = 0; i < path.size();i ++) {
					temp.add(path.get(i));
				}
				ans.add(temp);
				return;
			}
			if (root.left != null) {
				path.add(root.val);
				search(root, sum, now + root.val);
				path.remove(path.size() - 1);
			}
			if (root.right != null) {
				path.add(root.val);
				search(root, sum, now + root.val);
				path.remove(path.size() - 1);
			}
		}
	}
}

File: python/P113_PathSumII.py
#coding=utf-8

'''
    url: leetcode.com/problems/path-sum-ii
    @author:     zxwtry
    @email:      zxwtry@qq.com
    @date:       2017年4月30日
    @details:    Solution: 102ms 24.53%
'''

class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution(object):
    def search(self, n, s, c, a, t):
        if n == None: return
        t.append(n.val)
        if n.left == None and n.right == None:
            if s == c + n.val: a.append(list(t))
        self.search(n.left, s, c+n.val, a, t)
        self.search(n.right, s, c+n.val, a, t)
        t.pop()
        
    def pathSum(self, n, s):
        """
        :type n: TreeNode
        :type s: int
        :rtype: List[List[int]]
        """
        a, t = [], []
        self.search(n, s, 0, a, t)
        return a
        